### PR TITLE
Simplify setup stage of Cast operator

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

## Category:

**Refactoring**

## Description:

- Reduce CPU time by removing unnecessary use of SetupBlocks in Cast operator

### Affected modules and functionalities:

Cast op

### Tests:

- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3238

### CODE REVIEW

File: dali/kernels/common/cast.cuh [hunk 1]
```diff
@@ -16,58 +16,48 @@
...
-};
-
-template <typename OType, typename IType>
-__device__ __forceinline__ void CastKernelInternal(const CastSampleDesc& sample,
-                                                   unsigned block_start, unsigned block_end) {
-  auto *out = static_cast<OType *>(sample.output);
-  const auto *in = static_cast<const IType *>(sample.input);
-  for (unsigned x = threadIdx.x + block_start; x < block_end; x += blockDim.x) {
-    out[x] = ConvertSat<OType>(in[x]);
+inline __device__ uint32_t FindSampleIdx(const SampleDesc *samples,
```
Comment (mzient) [inline dali/kernels/common/cast.cuh:34]: :+1:

File: dali/operators/generic/cast.cu [hunk 2]
```diff
@@ -29,96 +28,46 @@ namespace dali {
...
-
  private:
-  using GpuBlockSetup = kernels::BlockSetup<1, -1>;
-  TensorListShape<1> collapsed_shape_;
-
-  GpuBlockSetup block_setup_;
-  std::vector<kernels::CastSampleDesc> samples_;
-
+  static constexpr int kBlockSize = 1024;
+  static constexpr int kLogicalBlockSize = 1024;
```
Comment (mzient) [inline dali/operators/generic/cast.cu:37]: Is it really the best thing? This looks rather suspicious - we run a binary search to copy one element per thread! This seems inefficient.

Comment (jantonguirao) [reply dali/operators/generic/cast.cu:37]: Not tuned yet. I'll update in a bit

File: dali/operators/generic/cast.cu [hunk 3]
```diff
@@ -29,96 +29,30 @@ namespace dali {
...
-      kernels::BinSearchCastKernel<OType, IType>
-          <<<grid_dim, block_dim, 0, ws.stream()>>>(samples_dev, params_dev,
-            num_samples, block_volume_scale);
+  TYPE_SWITCH(output.type(), type2id, Out, CAST_ALLOWED_TYPES, (
+    TYPE_SWITCH(input.type(), type2id, In, CAST_ALLOWED_TYPES, (
+      auto kernel = kernels::cast::CastGPU<Out, In>{};
+      kernel.Run(ctx, flatten(view<Out>(output)), flatten(view<const In>(input)));
     ), DALI_FAIL(make_string("Invalid input type: ", input.type())););  // NOLINT(whitespace/parens)
   ), DALI_FAIL(make_string("Invalid output type: ", output.type())););  // NOLINT(whitespace/parens)
+  CUDA_CALL(cudaGetLastError());
```
Comment (mzient) [inline dali/operators/generic/cast.cu:55]: This should go to the kernel's Run method.

**Review by mzient — APPROVED** [review]

**Review by banasraf — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Pull request merged**